torch/cuda/amp/grad_scaler.py

```diff
@@ -186,9 +186,20 @@ def _unscale_grads_(self, optimizer, inv_scale, found_inf, allow_fp16):
                     if (not allow_fp16) and param.grad.dtype == torch.float16:
                         raise ValueError("Attempting to unscale FP16 gradients.")
                     else:
-                        torch._amp_non_finite_check_and_unscale_(param.grad,
-                                                                 per_device_found_inf.get(param.grad.device),
-                                                                 per_device_inv_scale.get(param.grad.device))
+                        g = param.grad
+                        to_unscale = g._values() if g.is_sparse else g
+                        if g.is_sparse and g.dtype is torch.float16 and not g.is_coalesced():
+                            # is_coalesced() = False means the sparse grad has values with duplicate indices.
+                            # coalesce() deduplicates indices and adds all values that have the same index.
+                            # For scaled fp16 values, there's a good chance coalescing will cause overflow,
+                            # so we should double check the coalesced _values().
+                            torch._amp_non_finite_check_and_unscale_(g.coalesce()._values(),
```

Comment:
should you just replace g with its coalesced version in this case, and not do the check twice?

Reply:
That's a good idea, my original thinking was I don't want to replace param.grad with a new reference if I don't have to.  If you think that's ok it's a simple change.

Reply:
https://github.com/pytorch/pytorch/pull/36786/commits/0c7bc4f5b8e48cbfc6a5db6e29bb4ab26ebff846 replaces param.grad with the coalesced and unscaled version if param.grad was fp16 and uncoalesced.